src/codemirror/widget.ts

```diff
@@ -39,48 +54,171 @@ class CodeMirrorWidget extends Widget {
+    this._static.dispose();
+    this._live.dispose();
     super.dispose();
   }
 
   /**
    * Get the editor wrapped by the widget.
+   */
+  get editor(): CodeMirror.Editor {
+    return this._live.editor;
+  }
+
+  /**
+   * Handle `'activate-request'` messages.
+   */
+  protected onActivateRequest(msg: Message): void {
+    this._activate();
+  }
+
+  /**
+   * Handle the DOM events for the widget.
+   *
+   * @param event - The DOM event sent to the widget.
    *
    * #### Notes
-   * This is a ready-only property.
+   * This method implements the DOM `EventListener` interface and is
+   * called in response to events on the notebook panel's node. It should
+   * not be called directly by user code.
    */
-   get editor(): CodeMirror.Editor {
-     return this._editor;
-   }
+  handleEvent(event: Event): void {
+    switch (event.type) {
+    case 'mousedown':
+      this._evtMouseDown(event as MouseEvent);
+      break;
+    case 'focus':
+      this._evtFocus(event as MouseEvent);
+      break;
+    case 'blur':
+      this._evtBlur(event as MouseEvent);
+      break;
+    default:
+      break;
+    }
+  }
 
   /**
    * A message handler invoked on an `'after-attach'` message.
    */
   protected onAfterAttach(msg: Message): void {
-    if (!this.isVisible) {
-      this._needsRefresh = true;
+    this.node.addEventListener('mousedown', this);
+    this.node.addEventListener('focus', this, true);
+    this.node.addEventListener('blur', this, true);
+  }
+
+  /**
+   * Handle `before_detach` messages for the widget.
+   */
+  protected onBeforeDetach(msg: Message): void {
+    this.node.removeEventListener('mousedown', this);
+    this.node.removeEventListener('focus', this, true);
+    this.node.removeEventListener('blur', this, true);
+  }
+
+  /**
+   * Handle `mousedown` events for the widget.
+   */
+  private _evtMouseDown(event: MouseEvent): void {
+    if (this._live.isVisible) {
       return;
     }
-    this._editor.refresh();
-    this._needsRefresh = false;
+    this._lastMouseDown = event;
+  }
+
+  /**
+   * Handle `focus` events for the widget.
+   */
+  private _evtFocus(event: MouseEvent): void {
+    this._activate();
+  }
+
+  /**
+   * Handle `blur` events for the widget.
+   */
+  private _evtBlur(event: MouseEvent): void {
+    this._lastMouseDown = null;
+    if (this.node.contains(event.relatedTarget as HTMLElement)) {
+      return;
+    }
+    this._live.hide();
+    this._static.show();
+  }
+
+  /**
+   * Handle an activation message or a focus event.
+   */
+  private _activate(): void {
+    let editor = this.editor;
+    if (editor.getOption('readOnly') !== false) {
+      this._lastMouseDown = null;
+      return;
+    }
+    this._static.hide();
+    this._live.show();
+    if (this._lastMouseDown) {
+      let x = this._lastMouseDown.clientX;
+      let y = this._lastMouseDown.clientY;
+      let pos = editor.coordsChar({ left: x, top: y });
+      editor.getDoc().setCursor(pos);
```

Comment:
> We can't use the MouseEvent constructor in IE 

Oh, good to know!

> and it would still require some hackery of target, this is straightforward and uses the existing API.

Definitely. +1.

Thanks!